app/src/org/commcare/views/widgets/MediaWidget.java

```diff
@@ -162,4 +180,117 @@ public void setFocus(Context context) {
+            Log.i(TAG, "Inserting media returned uri = " + mediaUriString);
+            showToast("form.attachment.success");
+        } else {
+            Log.e(TAG, "Inserting media file FAILED");
+        }
+
+        mBinaryName = newMedia.getName();
+    }
+
+    /**
+     * If file is chosen by user, the file selection intent will return an URI
+     * If file is auto-selected after recording_fragment, then the recordingfragment will provide a string file path
+     * Set value of customFileTag if the file is a recent recording from the RecordingFragment
+     */
+    protected String createFilePath(Object binaryuri) {
+        String path = "";
+        destMediaPath = "";
+        if (binaryuri instanceof Uri) {
+            try {
+                path = UriToFilePath.getPathFromUri(CommCareApplication.instance(),
+                        (Uri)binaryuri);
+            } catch (UriToFilePath.NoDataColumnForUriException e) {
+                // Need to make a copy of file using uri, so might as well copy to final destination path directly
+                InputStream inputStream;
+                try {
+                    inputStream = getContext().getContentResolver().openInputStream((Uri)binaryuri);
+                } catch (FileNotFoundException e1) {
+                    showToast("form.attachment.notfound");
+                    e1.printStackTrace();
+                    return "";
+                }
+
+                recordedFileName = FileUtil.getFileName(((Uri)binaryuri).getPath());
+                destMediaPath = mInstanceFolder + System.currentTimeMillis() + FileUtil.getExtension(((Uri)binaryuri).getPath());
+
+                try {
+                    FileUtil.copyFile(inputStream, new File(destMediaPath));
+                } catch (IOException e1) {
+                    e1.printStackTrace();
+                    showToast("form.attachment.copy.fail");
+                    return "";
+                }
+                path = destMediaPath;
+            }
+            customFileTag = "";
+        } else {
+            path = (String)binaryuri;
+            customFileTag = CUSTOM_TAG;
+        }
+        return path;
+    }
+
+    protected void playMedia(String mediaType) {
+        Intent i = new Intent("android.intent.action.VIEW");
+        File mediaFile = new File(mInstanceFolder + mBinaryName);
+        Uri mediaUri = FileUtil.getUriForExternalFile(getContext(), mediaFile);
+        i.setDataAndType(mediaUri, mediaType);
+
+        UriToFilePath.grantPermissionForUri(getContext(), i, mediaUri, Intent.FLAG_GRANT_READ_URI_PERMISSION);
+        try {
+            getContext().startActivity(i);
+        } catch (ActivityNotFoundException e) {
+            Toast.makeText(getContext(),
+                    StringUtils.getStringSpannableRobust(getContext(),
+                            R.string.activity_not_found,
+                            "play media"),
```

Comment:
I think it's valuable to specify which type of media activity we weren't able to find.